packages/block-library/src/legacy-widget/edit/form.js

```diff
@@ -154,26 +127,109 @@ function useForm( { id, idBase, instance, setInstance } ) {
 	return { html, setFormData };
 }
 
-function serializeForm( form ) {
-	return new window.URLSearchParams(
-		Array.from( new window.FormData( form ) )
-	).toString();
-}
+function Control( { id, idBase, html, onChange, onSave } ) {
+	const controlRef = useRef();
+	const formRef = useRef();
 
-const ObservableForm = forwardRef( ( { onChange, ...props }, ref ) => {
-	// React won't call the form's onChange handler because it doesn't know
-	// about the <input>s that we add using dangerouslySetInnerHTML. We work
-	// around this by not using React's event system.
+	// Trigger widget-added when widget is ready and widget-updated when widget
+	// changes. This event is what widgets' scripts use to initialize, attach
+	// events, etc.
+	const hasBeenAdded = useRef( false );
+	useEffect( () => {
+		if ( html ) {
+			$( document ).trigger(
```

Comment:
So you only need to hook into the jQuery event system, right?

Reply:
@youknowriad or @adamsilverstein, do you remember how we did something similar to integrate the block editor with the heartbeat handler for the post lock overlay?

Reply:
I believe Core just "duplicates" its event bus by always [dispatching an event through both](https://github.com/WordPress/wordpress-develop/blob/master/src/js/_enqueues/wp/heartbeat.js#L396-L397) jQuery and `@wordpress/hooks`:

```js
$document.trigger( 'heartbeat-send', [ heartbeatData ] );
wp.hooks.doAction( 'heartbeat.send', heartbeatData );
```

That way, Gutenberg and other modern consumers can just use `addAction`:

https://github.com/WordPress/gutenberg/blob/a0bba9ea67e8985b686cf8d55eccba42f9dcdfce/packages/editor/src/components/post-locked-modal/index.js#L129-L133

***

Short of doing the same in Core for widget events — and at the risk of asking a stupid question — could we get away with not importing jQuery and making these React effects conditional upon the availability of jQuery? That is:

```js
useEffect( () => {
  if ( ! window.jQuery ) return;

  const { jQuery: $ } = window;

  if ( html ) $( document ).trigger( ... );
} );
```

Reply:
Interesting! We can't realistically use this solution here because it would require updating every third party widget that uses `$.trigger()` to also perform a `doAction`.

I think accessing jQuery from `window` and falling back gracefully if it does not exist seems like the best compromise here.

Reply:
I did this in 216e286.